tests/testsuite/future_incompat_report.rs

```diff
@@ -0,0 +1,165 @@
+        .with_stderr_contains("error: the `--future-incompat-report` flag is unstable[..]")
+        .with_status(101)
+        .run();
+
+    p.cargo("build --future-incompat-report -Z unstable-options")
+        .masquerade_as_nightly_cargo()
+        .with_stderr_contains(
+            "error: Usage of `--future-incompat-report` requires `-Z future-incompat-report`",
+        )
+        .with_status(101)
+        .run();
+
+    p.cargo("describe-future-incompatibilities --id foo")
+        .with_stderr_contains(
+            "error: `cargo describe-future-incompatibilities` can only be used on the nightly channel"
+        )
+        .with_status(101)
+        .run();
+}
+
+#[cargo_test]
+fn test_single_crate() {
+    if !is_nightly() {
+        return;
+    }
+
+    let p = project()
+        .file("Cargo.toml", &basic_manifest("foo", "0.0.0"))
+        .file("src/main.rs", "fn main() { [true].into_iter(); }")
+        .build();
+
+    for command in &["build", "check", "rustc"] {
+        p.cargo(command).arg("-Zfuture-incompat-report")
+            .masquerade_as_nightly_cargo()
+            .with_stderr_contains("  = note: `#[warn(array_into_iter)]` on by default")
+            .with_stderr_contains("warning: the following crates contain code that will be rejected by a future version of Rust: foo v0.0.0 [..]")
+            .with_stderr_does_not_contain("[..] future incompatibility lints:")
+            .run();
+
+        p.cargo(command).arg("-Zfuture-incompat-report").arg("-Zunstable-options").arg("--future-incompat-report")
+            .masquerade_as_nightly_cargo()
+            .with_stderr_contains("  = note: `#[warn(array_into_iter)]` on by default")
+            .with_stderr_contains("warning: the following crates contain code that will be rejected by a future version of Rust: foo v0.0.0 [..]")
+            .with_stderr_contains("The crate `foo v0.0.0 ([..])` currently triggers the following future incompatibility lints:")
+            .run();
+    }
+}
+
+#[cargo_test]
+fn test_multi_crate() {
+    if !is_nightly() {
+        return;
+    }
+
+    Package::new("first-dep", "0.0.1")
+        .file("src/lib.rs", "fn foo() { [25].into_iter(); }")
+        .publish();
+    Package::new("second-dep", "0.0.2")
+        .file("src/lib.rs", "fn foo() { ['a'].into_iter(); }")
+        .publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.0"
+
+                [dependencies]
+                first-dep = "*"
+                second-dep = "*"
+              "#,
+        )
+        .file("src/main.rs", "fn main() {}")
+        .build();
+
+    for command in &["build", "check", "rustc"] {
+        p.cargo(&format!("{} -Z future-incompat-report", command))
+            .masquerade_as_nightly_cargo()
+            .with_stderr_does_not_contain("  = note: `#[warn(array_into_iter)]` on by default")
```

Comment:
I'm pretty uncomfortable with all of these `with_stderr_does_not_contain` calls.  That function is pretty fragile, because any change in output means it will no longer be validating the behavior.  And rustc's output is prone to change frequently, and there will be no way to know when it changes.

Usually to make it more reliable, I would have a test that has `with_stderr_contains("[..]example[..]")`, and then a second test that validates the opposite with the exact same string with `with_stderr_does_not_contain("[..]example[..]")`.

And, in general, when checking rustc's output, I try to use wildcards as much as possible.  If someone wants to change the output, and it breaks cargo's testsuite, then it requires a hassle to update cargo to somehow work with the change.

In these cases, it might be less fragile to match against things like `"[..]array_into_iter[..]"`.